plugins/modem-manager/fu-mbim-qdu-updater.c

```diff
@@ -0,0 +1,461 @@
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	total_sending_numbers = (g_bytes_get_size (ctx->blob)) / (out_max_transfer_size);
+	if ((g_bytes_get_size (ctx->blob)) % (out_max_transfer_size) != 0)
+		total_sending_numbers ++;
+
+	ctx->chunks = fu_chunk_array_new_from_bytes (ctx->blob,
+						     0x00,	/* start addr */
+						     0x00,	/* page_sz */
+						     out_max_transfer_size);
+	chk = g_ptr_array_index (ctx->chunks, 0);
+	request = mbim_message_qdu_file_write_set_new (fu_chunk_get_data_sz (chk), (const guint8 *)fu_chunk_get_data (chk), NULL);
+	mbim_device_command (ctx->mbim_device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_write_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_session_ready (MbimDevice *device, GAsyncResult *res, gpointer user_data)
+{
+	g_autoptr(MbimMessage) response = NULL;
+	MbimQduSessionType     out_current_session_type;
+	MbimQduSessionStatus   out_current_session_status;
+	MbimQduSessionType     out_last_session_type;
+	MbimQduSessionResult   out_last_session_result;
+	guint32                out_last_session_error_offset;
+	guint32                out_last_session_error_size;
+	g_autoptr(GError)      error = NULL;
+	g_autoptr(MbimMessage) request = NULL;
+	WriteContext *ctx = user_data;
+
+	response = mbim_device_command_finish (device, res, &error);
+	if (!response || !mbim_message_response_get_result (response, MBIM_MESSAGE_TYPE_COMMAND_DONE, &error)) {
+		g_debug ("error: operation failed: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (!mbim_message_qdu_update_session_response_parse (response,
+							     &out_current_session_type,
+							     &out_current_session_status,
+							     &out_last_session_type,
+							     &out_last_session_result,
+							     &out_last_session_error_offset,
+							     &out_last_session_error_size,
+							     &error)) {
+		g_debug ("error: couldn't parse response message: %s\n", error->message);
+		g_main_loop_quit (ctx->mainloop);
+		return;
+	}
+
+	if (response)
+		mbim_message_unref (response);
+
+	g_debug ("[%s] Successfully request modem to update session", mbim_device_get_path_display (device));
+
+	request = mbim_message_qdu_file_open_set_new (0, g_bytes_get_size (ctx->blob), NULL);
+	mbim_device_command (device,
+			     request,
+			     10,
+			     ctx->cancellable,
+			     (GAsyncReadyCallback)fu_mbim_qdu_updater_file_open_ready,
+			     ctx);
+	mbim_message_unref (request);
+}
+
+static void
+fu_mbim_qdu_updater_set_update_session (MbimDevice *device, WriteContext *ctx)
+{
+  	g_autoptr(MbimMessage) request = NULL;
+
+	request = (mbim_message_qdu_update_session_set_new (0, 1, NULL));
+	if (!request)
+		g_debug ("error: couldn't create request: \n");
+	else
+		g_debug ("Create request");
```

Comment:
if this is useful debugging information, why not just collapse it to something like this:
```
g_debug ("request creation: %d", request != NULL);
```

Reply:
Thanks for your suggestions, it's better to collapse it as you say.

Reply:
There is really no need for this, you can assume `mbim_message_qdu_update_session_set_new()` will never fail and simplify everything :)

Reply:
No no, you must use the `request`. What I mean is that `request` will always be non-NULL, so you don't need to `if (!request)`